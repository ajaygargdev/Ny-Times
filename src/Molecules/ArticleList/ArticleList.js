import React, { useMemo, useState } from "react";
import { Container } from "@mui/material";
import ArticleCard from "../ArticleCard/ArticleCard";
import { useSelector } from "react-redux";
import ModelArticle from "../ModelArticle/ModelArticle";

const ArticlesList = () => {
  const articles = useSelector((state) => state.root.articles);
  const currentPage = useSelector((state) => state.root.currentPage);
  const [currentArticle, setCurrentArtical] = useState(null);

  const articlesList = useMemo(
    () => articles[currentPage] || [],
    [currentPage, articles],
  );

  const handleClick = (article) => setCurrentArtical(article);

  const handleClose = () => setCurrentArtical(null);

  return (
    <>
      <Container
        data-testid="artical-list"
        sx={{
          "flex-wrap": "wrap",
          display: "flex",
          gap: "20px",
          justifyContent: "center",
        }}
      >
        {Array.isArray(articlesList) &&
          articlesList.map((article) => (
            <ArticleCard
              key={articles.asset_id}
              article={article}
              handleClick={() => handleClick(article)}
            />
          ))}
      </Container>
      {currentArticle && (
        <ModelArticle article={currentArticle} handleClose={handleClose} />
      )}
    </>
  );
};

export default ArticlesList;
