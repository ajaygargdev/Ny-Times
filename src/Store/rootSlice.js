import { createSlice, createAsyncThunk } from "@reduxjs/toolkit";
import { getArticles } from "../Services";

export const fetchArticles = createAsyncThunk(
  "root/fetchArticles",
  async (page = 1) => {
    console.log("-----test----");
    const response = await getArticles(page);
    if (!response.success) {
      throw new Error("Network response was not ok");
    }
    return { ...response.data, page };
  },
);

const initialState = {
  articles: [],
  copyright: "",
  isLoading: false,
  error: null,
  currentPage: "1",
};

export const rootSlice = createSlice({
  name: "root",
  initialState,
  reducers: {
    changePage: (state, payload) => {
      state.currentPage = payload + "";
    },
  },
  extraReducers: (builder) => {
    builder
      .addCase(fetchArticles.pending, (state) => {
        state.isLoading = true;
        state.error = null;
      })
      .addCase(fetchArticles.fulfilled, (state, action) => {
        state.isLoading = false;
        state.articles = {
          ...state.articles,
          [action.payload.page]: action.payload?.results || [],
        };
        state.copyright = action.payload.copyright;
      })
      .addCase(fetchArticles.rejected, (state, action) => {
        state.isLoading = false;
        state.error = action.error;
      });
  },
});

export const { changePage } = rootSlice.actions;

export default rootSlice.reducer;
