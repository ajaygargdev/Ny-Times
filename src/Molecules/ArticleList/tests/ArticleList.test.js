/**
 * @jest-environment jsdom
 */

import React from "react";
import { render, screen, fireEvent, waitFor } from "@testing-library/react";
import ArticleList from "../ArticleList";
import { mockData } from "./ArticalList.mock";
import { Provider } from "react-redux";
import { store } from "../../../Store/Store";
import configureStore from "redux-mock-store";
import thunk from "redux-thunk";

const middleware = [thunk];
const mockStore = configureStore({ middleware });

// jest.mock("react-redux", () => ({
//   ...jest.requireActual("react-redux"),
//   useSelector: jest.fn(),
//   useDispatch: jest.fn(),
// }));

// require("react-redux").useSelector.mockImplementation((callback) =>
//   callback({
//     root: {
//       articles: { "1": mockData.results[0] },
//       currentPage: "1",
//     },
//   }),
// );

describe("ArticalList Component", () => {
  let store;

  beforeEach(() => {
    store = mockStore({
      root: {
        articles: { "1": mockData.results[0] },
        currentPage: "1",
      },
    });
  });

  afterAll(() => {
    jest.clearAllMocks();
  });

  test("renders without crashing", () => {

    const { container } = render(
      <Provider store={store}>
        <ArticleList />
      </Provider>,
    );
    expect(container).toMatchSnapshot();
  });

  test("fire click on component", async () => {
    // require("react-redux").useSelector.mockImplementation((callback) =>
    //   callback({
    //     root: {
    //       articles: { "1": mockData.results[0] },
    //       currentPage: "1",
    //     },
    //   }),
    // );
    const {container,getByText} = await render(
      <Provider store={store}>
        <ArticleList />
      </Provider>,
    );
    await waitFor(async ()=>{
      const node =getByText(/The Heart-Healthy/i);
      await fireEvent.click(node);
      expect(container).toMatchSnapshot();  
    })
    // await new Promise((resolve) => setTimeout(resolve, 500));
    // console.log(screen.debug());
    // const node =await getByText(/The Heart-Healthy/i);
    // fireEvent.click(node);
    // expect(container).toMatchSnapshot();
  });
});
