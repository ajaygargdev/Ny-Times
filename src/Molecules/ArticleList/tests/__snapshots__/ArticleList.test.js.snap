// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ArticalList Component renders without crashing 1`] = `
<div>
  <div
    class="MuiContainer-root MuiContainer-maxWidthLg css-1lgbfpa-MuiContainer-root"
  />
</div>
`;
